code/ship/ship.cpp

```diff
@@ -8856,10 +8856,67 @@ static void lethality_decay(ai_info *aip)
 #endif
 }
 
-void ship_process_pre(object *objp, float frametime)
+void ship_process_pre(object *obj, float frametime)
 {
-	if ( (objp == NULL) || !frametime )
+	// If Ai_before_physics is false everything following is evaluated in ship_process_post()
+	if ( (obj == nullptr) || !frametime || !Ai_before_physics)
 		return;
+
+	int	num;
+	ship* shipp;
+
+	if (obj->type != OBJ_SHIP) {
+		nprintf(("Network", "Ignoring non-ship object in ship_process_pre()\n"));
+		return;
+	}
+
+	num = obj->instance;
+	Assert(num >= 0 && num < MAX_SHIPS);
+	Assert(obj->type == OBJ_SHIP);
+	Assert(Ships[num].objnum == OBJ_INDEX(obj));
+
+	shipp = &Ships[num];
+
+	//rotate player subobjects since its processed by the ai functions
+	// AL 2-19-98: Fire turret for player if it exists
+	//WMC - changed this to call process_subobjects
+	if ((obj->flags[Object::Object_Flags::Player_ship]) && !Player_use_ai)
+	{
+		ai_info* aip = &Ai_info[Ships[obj->instance].ai_index];
+		if (aip->ai_flags[AI::AI_Flags::Being_repaired, AI::AI_Flags::Awaiting_repair])
+		{
+			if (aip->support_ship_objnum >= 0)
+			{
+				if (vm_vec_dist_quick(&obj->pos, &Objects[aip->support_ship_objnum].pos) < (obj->radius + Objects[aip->support_ship_objnum].radius) * 1.25f)
+					return;
+			}
+		}
+		if (!shipp->flags[Ship::Ship_Flags::Rotators_locked])
+			process_subobjects(OBJ_INDEX(obj));
+	}
+
+	if (obj == Player_obj) {
+		ship_check_player_distance();
+	}
+
+	// update ship lethality
+	if (Ships[num].ai_index >= 0) {
+		if (!physics_paused && !ai_paused) {
+			lethality_decay(&Ai_info[Ships[num].ai_index]);
+		}
+	}
+
+	// if the ship is an observer ship don't need to do AI
+	if (obj->type == OBJ_OBSERVER) {
+		return;
+	}
+
+	// Goober5000 - player may want to use AI
+	if ((Ships[num].ai_index >= 0) && (!(obj->flags[Object::Object_Flags::Player_ship]) || Player_use_ai)) {
+		if (!physics_paused && !ai_paused) {
+			ai_process(obj, Ships[num].ai_index, frametime);
+		}
+	}
```

Comment:
Am I right in thinking that this is a copy of code from elsewhere?  If they have the same exact code, it may be helpful to put both instances of them in a helper function so that changes to one do not have to be duplicated somewhere else.  Unless there are plans for them to diverge at some point.

But maybe someone with more experience can weigh in on that.